### PR TITLE
Avoid subregisters

### PR DESCRIPTION
**Pull request by mellowCS**
This PR contains the processing of Sub-Registers. They are now casted into Subpieces of their corresponding Base-Registers and either zero extended or pieced together depending on the following instructions.

### CODE REVIEW

File: cwe_checker_rs/src/pcode/term.rs [hunk 1]
```diff
@@ -410,10 +412,93 @@ pub struct Project {
...
+        let mut program: Term<IrProgram> = Term {
             tid: project.program.tid,
             term: project.program.term.into(),
         };
+        let register_map: HashMap<&String, &RegisterProperties> = project
+            .register_properties
+            .iter()
+            .map(|p| (&p.register, p))
+            .collect();
+        let mut zero_extend_tids: HashMap<String, Tid> = HashMap::new();
```
Comment (Enkelmann) [inline cwe_checker_rs/src/pcode/term.rs:424]: It seems like you want to use a `HashSet<Tid>` here instead of a HashMap.

File: cwe_checker_rs/src/pcode/term.rs [hunk 2]
```diff
@@ -410,10 +412,93 @@ pub struct Project {
...
+                                Some(var),
+                                &register_map,
+                                peeked_def,
+                            );
+                            match zero_tid {
+                                Some(tid) => {
+                                    zero_extend_tids.insert(format!("{}", tid), tid);
+                                }
+                                _ => (),
+                            }
```
Comment (Enkelmann) [inline cwe_checker_rs/src/pcode/term.rs:444]: You can combine the `let` with the `match` to an `if let Some(zero_tid) = ...` to make the code more concise.

File: cwe_checker_rs/src/pcode/term.rs [hunk 3]
```diff
@@ -410,10 +412,93 @@ pub struct Project {
...
```
Comment (Enkelmann) [inline cwe_checker_rs/src/pcode/term.rs:457]: You can combine the `let` with the `match` to an `if let Some(zero_tid) = ...` to make the code more concise.

File: cwe_checker_rs/src/pcode/term.rs [hunk 4]
```diff
@@ -410,10 +412,93 @@ pub struct Project {
...
+                            value.process_sub_registers_if_necessary(
+                                None,
+                                &register_map,
+                                peeked_def,
+                            );
+                        }
+                    }
+                }
+                let mut jmp_iter = blk.term.jmps.iter_mut();
+                while let Some(jmp) = jmp_iter.next() {
```
Comment (Enkelmann) [inline cwe_checker_rs/src/pcode/term.rs:474]: use `for jmp in blk.term.jmps.iter_mut()` here.

**Review by Enkelmann — CHANGES_REQUESTED** [review]
Clippy also had some comments. Make sure to run `make clean` before running `make codestyle-check`. Else some Clippy warnings may be suppressed. Normally Travis would point this out, but I still have to find out why it is not running right now.

File: cwe_checker_rs/src/intermediate_representation/expression.rs [hunk 5]
```diff
@@ -132,6 +134,219 @@ impl Expression {
...
+    ) -> Option<Tid> {
+        let mut output_base_size: Option<ByteSize> = None;
+        let mut peek_is_zero_extension: bool = false;
+        let mut output_base_register: Option<&&RegisterProperties> = None;
+        let mut output_sub_register: Option<&RegisterProperties> = None;
+        let mut zero_extend_tid: Option<Tid> = None;
+
+        if let Some(output_value) = output {
+            if let Some(register) = register_map.get(&output_value.name) {
+                if *register.register != *register.base_register {
```
[the remark this answers is not in the document]
Comment (Enkelmann) [reply cwe_checker_rs/src/intermediate_representation/expression.rs:156]: Oops, my mistake, forget this comment. ;-)

File: cwe_checker_rs/src/intermediate_representation/term.rs [hunk 6]
```diff
@@ -68,6 +68,33 @@ pub enum Def {
...
 }
 
+impl Term<Def> {
+    /// This function checks whether the following instruction
+    /// is a zero extension of the currently overwritten sub register and
+    /// if so, returns its TID
+    pub fn check_for_zero_extension(&self, output_name: String) -> Option<Tid> {
+        match &self.term {
+            Def::Assign { var, value } => {
+                if output_name == var.name {
```
Comment (Enkelmann) [inline cwe_checker_rs/src/intermediate_representation/term.rs:78]: I think you can combine that to `Def::Assign {var, value } if output_name == var.name => {` to get rid of a nesting level here.

**Review by Enkelmann — CHANGES_REQUESTED** [review]
You missed two issues from the first review. ;-)

File: cwe_checker_rs/src/intermediate_representation/term.rs [hunk 7]
```diff
@@ -68,6 +68,33 @@ pub enum Def {
...
+                    match value {
+                        Expression::Cast { op, arg, .. } => match op {
+                            CastOpType::IntZExt => match **arg {
+                                Expression::Var(_) => Some(self.tid.clone()),
```
[the remark this answers is not in the document]
Comment (mellowCS) [reply cwe_checker_rs/src/intermediate_representation/term.rs:82]: I only check if the output register of the zero extension is the correct one. Overwriting the full register with a different input here would not make sense after overwriting the sub register in the instruction before. I mean I might happen but it would not make sense.

File: cwe_checker_rs/src/intermediate_representation/term.rs [hunk 8]
```diff
@@ -68,6 +68,33 @@ pub enum Def {
     Assign { var: Variable, value: Expression },
 }
 
+impl Term<Def> {
+    /// This function checks whether the following instruction
+    /// is a zero extension of the currently overwritten sub register and
+    /// if so, returns its TID
+    pub fn check_for_zero_extension(&self, output_name: String) -> Option<Tid> {
```
[the remark this answers is not in the document]
Comment (mellowCS) [reply cwe_checker_rs/src/intermediate_representation/term.rs:75]: True, I'll change that!

File: cwe_checker_rs/src/intermediate_representation/term.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (mellowCS) [reply cwe_checker_rs/src/intermediate_representation/term.rs:82]: I have just implemented the check just in case.

[the remark this answers is not in the document]
Comment (Enkelmann) [reply cwe_checker_rs/src/intermediate_representation/term.rs]: It may not make sense, but that doesn't mean that it cannot happen. E.g. in obfuscated binaries you have to expect operations that are only there to confuse. So we should make sure that such corner cases do not lead to errors.

**Review by Enkelmann — APPROVED** [review]
Looks good!
All tests succeeded on my PC, so I will merge it despite Travis not running. Hopefully, we get Travis back next week.
